### PR TITLE
Create gcs.json and parse primary key from GCS files

### PR DESCRIPTION
**Pull request by melissachang**
This PR reads files from GCS and parses primary key from file path. Nothing is indexed yet; that will be in next PR.

Indexer output:
```
indexer_1        | Identified primary key NA12877 from gs://genomics-public-data/platinum-genomes/bam/NA12877_S1.bam
indexer_1        | Identified primary key NA12877 from gs://genomics-public-data/platinum-genomes/bam/NA12877_S1.bam.bai
indexer_1        | Identified primary key NA12878 from gs://genomics-public-data/platinum-genomes/bam/NA12878_S1.bam
indexer_1        | Identified primary key NA12878 from gs://genomics-public-data/platinum-genomes/bam/NA12878_S1.bam.bai
indexer_1        | Identified primary key NA12889 from gs://genomics-public-data/platinum-genomes/bam/NA12889_S1.bam
indexer_1        | Identified primary key NA12889 from gs://genomics-public-data/platinum-genomes/bam/NA12889_S1.bam.bai
indexer_1        | Identified primary key NA12890 from gs://genomics-public-data/platinum-genomes/bam/NA12890_S1.bam
indexer_1        | Identified primary key NA12890 from gs://genomics-public-data/platinum-genomes/bam/NA12890_S1.bam.bai
indexer_1        | Identified primary key NA12891 from gs://genomics-public-data/platinum-genomes/bam/NA12891_S1.bam
indexer_1        | Identified primary key NA12891 from gs://genomics-public-data/platinum-genomes/bam/NA12891_S1.bam.bai
indexer_1        | Identified primary key NA12892 from gs://genomics-public-data/platinum-genomes/bam/NA12892_S1.bam
indexer_1        | Identified primary key NA12892 from gs://genomics-public-data/platinum-genomes/bam/NA12892_S1.bam.bai
```

### CODE REVIEW

File: gcs/indexer.py [hunk 1]
```diff
@@ -39,14 +41,53 @@ def parse_args():
...
+        else:
+            raise ValueError('Could not find primary key in %s' % obj_path)
+
+
 def main():
     args = parse_args()
 
+    # Read dataset config files
     index_name = indexer_util.get_index_name(args.dataset_config_dir)
+    gcs_json = os.path.join(args.dataset_config_dir, 'gcs.json')
```
Comment (alahwa) [inline gcs/indexer.py]: `gcs_json` sounds like a JSON object or a string containing json. `gcs_config_path` would be clearer

File: dataset_config/platinum_genomes/gcs.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Fields for GCS indexer
+{
+  "gcs_paths": [
+    // GCS path with PRIMARY_KEY. There is an implied .+ at the end.
+    // Example: gs://genomics-public-data/platinum-genomes/bam/PRIMARY_KEY_
+    // For more information about primary key, see
+    // https://github.com/DataBiosphere/data-explorer-indexers#overview
+    "gs://genomics-public-data/platinum-genomes/bam/PRIMARY_KEY_"
```
Comment (alahwa) [inline dataset_config/platinum_genomes/gcs.json]: I'm confused about what PRIMARY_KEY is doing here. Is this config asking me to specify my path as `gs://test-data-explorer/directory/path/PRIMARY_KEY_participant_id`?

File: gcs/indexer.py (hunk 1 above)
Comment (melissachang) [reply gcs/indexer.py]: Done

File: dataset_config/platinum_genomes/gcs.json (hunk 2 above)
Comment (melissachang) [reply dataset_config/platinum_genomes/gcs.json]: Expanded comments. Does it make sense now?

**Comment by alahwa** [comment]
git push? :)

File: dataset_config/platinum_genomes/gcs.json (hunk 2 above)
Comment (alahwa) [reply dataset_config/platinum_genomes/gcs.json]: Thanks, I see what you're doing now.

Under this design, we assume that the PRIMARY_KEY is present in the file path, and also consistently placed at the beginning of either a subdirectory or file name. Both are pretty heavy assumptions. For example, I couldn't use this indexer if my GCS objects looked something like `20180728_NA12877_S1.bam` where `NA12877` correlates to participant_ids. These assumptions might hold up for AMP-PD as it currently is, but are pretty restrictive beyond that. Have you considered any alternatives?

Comment (alahwa) [reply dataset_config/platinum_genomes/gcs.json]: As discussed offline, let's keep tabs on how to make these assumptions less restrictive and document for now that GCS filepaths must contain primary_keys in a uniform manner that map to those used by other sources (i.e. bigquery)

Comment (melissachang) [reply dataset_config/platinum_genomes/gcs.json]: Done

File: gcs/README.md [hunk 3]
```diff
@@ -20,6 +20,17 @@ container from the [data-explorer repo](https://github.com/DataBiosphere/data-ex
 
 ### Index a custom dataset locally
 
+We assume that primary key appears in the GCS file path. For example, the file
```
Comment (alahwa) [inline gcs/README.md]: and that all files follow a single naming convention.

File: gcs/README.md [hunk 4]
```diff
@@ -20,6 +20,17 @@ container from the [data-explorer repo](https://github.com/DataBiosphere/data-ex
...
+`gs://genomics-public-data/platinum-genomes/bam/NA12890_S1.bam` has primary key
+`NA12890`. Primary key can appear anywhere in the path; being part of a
```
Comment (alahwa) [inline gcs/README.md]: can appear anywhere as long as it's consistent

File: dataset_config/platinum_genomes/gcs.json [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Fields for GCS indexer
+{
+  "gcs_paths": [
```
Comment (alahwa) [inline dataset_config/platinum_genomes/gcs.json]: Convention is to name this `gcs_pattern` indicating a regex string over `gcs_path` which indicates a static file path. I would also explicitly mention that regex is allowed here in addition to treating PRIMARY_KEY as a special substring.

I would prefer that users not have to supply regex, but this seems to be the easiest workaround for the time being.

**Pull request merged**